### PR TITLE
Use BigDecimal for Decimal feature value

### PR DESCRIPTION
**Pull request by ewanv**
Creates `Feature.Value.FloatingPoint` to hold `Double` values, and changes the value type of `Decimal` to `BigDecimal`. It may be better not to create a `FloatingPoint`, and allow `Decimal` to accept `Double` values, and convert them to `BigDecimal`. This change would then not break existing client code.

This PR also makes the `avg` aggregator use `BigDecimal` instead of `Double`.

### CODE REVIEW

File: core/src/main/scala/commbank/coppersmith/Feature.scala [hunk 1]
```diff
@@ -44,21 +44,25 @@ object Feature {
...
+    implicit def fromOInt(i: Option[Int]):                Integral      = Integral(i.map(_.toLong))
+    implicit def fromOLong(l: Option[Long]):              Integral      = Integral(l)
+    implicit def fromODouble(d: Option[Double]):          FloatingPoint = FloatingPoint(d)
+    implicit def fromOBigDecimal(bd: Option[BigDecimal]): Decimal       = Decimal(bd)
+    implicit def fromOString(s: Option[String]):          Str           = Str(s)
 
     implicit val intOrder: Order[Integral] = orderBy(_.value)
     implicit val decOrder: Order[Decimal] = orderBy(_.value)
+    implicit val fpOrder:  Order[FloatingPoint] = orderBy(_.value)
     implicit val strOrder: Order[Str] = orderBy(_.value)
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/Feature.scala]: Very minor: Consider vertically aligning on the `=` above.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 2]
```diff
@@ -112,31 +120,98 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
-  def uniqueCountBy[T](f : S => T): Aggregator[S, Set[T], Int]          = AggregationFeature.uniqueCountBy(f)
+  def size: Aggregator[S, Long, Long]                                          = Aggregator.size
+  def count(where: S => Boolean = _ => true): Aggregator[S, Long, Long]        = Aggregator.count(where)
+  def avg[V](v: S => BigDecimal): Aggregator[S, BigDAveragedValue, BigDecimal] = AggregationFeature.avg[S](v)
+  def max[V : Ordering](v: S => V): Aggregator[S, V, V]                        = AggregationFeature.max[S, V](v)
+  def min[V : Ordering](v: S => V): Aggregator[S, V, V]                        = AggregationFeature.min[S, V](v)
+  def maxBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V]        = AggregationFeature.maxBy[S, O, V](o)(v)
+  def minBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V]        = AggregationFeature.minBy[S, O, V](o)(v)
+  def sum[V : Monoid](v: S => V): Aggregator[S, V, V]                          = Aggregator.prepareMonoid(v)
+  def uniqueCountBy[T](f : S => T): Aggregator[S, Set[T], Int]                 = AggregationFeature.uniqueCountBy(f)
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: I'm not sure the value of vertical alignment above still outweighs the new longest line length. I'd consider wrapping after the `=` (similar to the `AggregationFeature` object below). We generally have a soft max line length of 100 characters.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 3]
```diff
@@ -112,31 +120,98 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
-  def avg[V](v: S => Double): Aggregator[S, AveragedValue, Double]      = AggregationFeature.avg[S](v)
-  def max[V : Ordering](v: S => V): Aggregator[S, V, V]                 = AggregationFeature.max[S, V](v)
-  def min[V : Ordering](v: S => V): Aggregator[S, V, V]                 = AggregationFeature.min[S, V](v)
-  def maxBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V] = AggregationFeature.maxBy[S, O, V](o)(v)
-  def minBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V] = AggregationFeature.minBy[S, O, V](o)(v)
-  def sum[V : Monoid]  (v: S => V): Aggregator[S, V, V]                 = Aggregator.prepareMonoid(v)
-  def uniqueCountBy[T](f : S => T): Aggregator[S, Set[T], Int]          = AggregationFeature.uniqueCountBy(f)
+  def size: Aggregator[S, Long, Long]                                          = Aggregator.size
+  def count(where: S => Boolean = _ => true): Aggregator[S, Long, Long]        = Aggregator.count(where)
+  def avg[V](v: S => BigDecimal): Aggregator[S, BigDAveragedValue, BigDecimal] = AggregationFeature.avg[S](v)
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: What's the reason for _replacing_ the `Double` implementation, as opposed to just _adding_ the new `BigDecimal` implementation? Is it not still desirable to do averages using doubles in some cases?

File: core/src/test/scala/commbank/coppersmith/Arbitraries.scala [hunk 4]
```diff
@@ -102,4 +108,6 @@ object Arbitraries {
 
   val arbNonEmptyAlphaStr: Gen[NonEmptyString] =
     for (h <- alphaChar; t <- alphaStr) yield nonEmptyString(h, t)
+
+  val hiveIdentifierGen: Gen[String] = arbNonEmptyAlphaStr.map(_.value.take(127))
```
Comment (dkristian) [inline core/src/test/scala/commbank/coppersmith/Arbitraries.scala]: Could you please add a comment about the need for `.take(127)`

File: core/src/test/scala/commbank/coppersmith/FeatureSpec.scala [hunk 5]
```diff
@@ -42,11 +42,13 @@ object MetadataSpec extends Specification with ScalaCheck { def is = s2"""
...
-        Metadata[Customer, Integral](namespace, name, desc, fType).valueType must_== IntegralType
+        Metadata[Customer, Integral]     (namespace, name, desc, fType).valueType must_== IntegralType
       case Decimal(_) =>
-        Metadata[Customer, Decimal] (namespace, name, desc, fType).valueType must_== DecimalType
+        Metadata[Customer, Decimal]      (namespace, name, desc, fType).valueType must_== DecimalType
+      case FloatingPoint(_) =>
+        Metadata[Customer, FloatingPoint](namespace, name, desc, fType).valueType must_== FloatingPointType
       case Str(_) =>
-        Metadata[Customer, Str]     (namespace, name, desc, fType).valueType.must_==(StringType)
+        Metadata[Customer, Str]          (namespace, name, desc, fType).valueType.must_==(StringType)
```
Comment (dkristian) [inline core/src/test/scala/commbank/coppersmith/FeatureSpec.scala]: I realise it was like that when you got here, but there's an inconsistency above wrt optional `.` and parens:
```scala
...valueType must_== ...
vs
...valuetype.must_==(...)
```
I'm not really fussed which style is used when using the `specs2` DSL, as long as it's consistent, especially within a single spec.

File: core/src/test/scala/commbank/coppersmith/QueryFeatureSpec.scala [hunk 6]
```diff
@@ -52,9 +52,9 @@ object QueryFeatureSetSpec extends Specification with ScalaCheck { def is = s2""
       queryFeature(name, humanDescription, condition)
     }
 
-    val youngHeight: Feature[Customer, Decimal] = feature("youngHeight", "Young Height", _.age < 18)
-    val midHeight:   Feature[Customer, Decimal] = feature("midHeight",  "Middle Height", c => Range(18 , 65).contains(c.age))
-    val oldHeight:   Feature[Customer, Decimal] = feature("oldHeight", "Old Height",  _.age >= 65)
+    val youngHeight: Feature[Customer, FloatingPoint] = feature("youngHeight", "Young Height",  _.age < 18)
+    val midHeight:   Feature[Customer, FloatingPoint] = feature("midHeight",   "Middle Height", c => Range(18 , 65).contains(c.age))
+    val oldHeight:   Feature[Customer, FloatingPoint] = feature("oldHeight",  "Old Height",     _.age >= 65)
```
Comment (dkristian) [inline core/src/test/scala/commbank/coppersmith/QueryFeatureSpec.scala]: Very minor: vertical alignment is slightly off above (probably obscured if the 2nd line wraps due to length). Consider adding a type alias to reduce line length:
```scala
    type CustFeat = Feature[Customer, FloatingPoint]
    val youngHeight: CustFeat = feature("youngHeight", "Young Height",  _.age < 18)
    val midHeight:   CustFeat = feature("midHeight",   "Middle Height", c => Range(18 , 65).contains(c.age))
    val oldHeight:   CustFeat = feature("oldHeight",   "Old Height",    _.age >= 65)
```

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala (hunk 3 above)
Comment (ewanv) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: If performance is similar, then enforcing the precision of `BigDecimal` may be preferred. I haven't done a performance test yet though.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 7]
```diff
@@ -112,31 +120,111 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
+  def min[V : Ordering](v: S => V): Aggregator[S, V, V]                        =
+    AggregationFeature.min[S, V](v)
+  def maxBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V]        =
+    AggregationFeature.maxBy[S, O, V](o)(v)
+  def minBy[O : Ordering, V](o: S => O)(v: S => V): Aggregator[S, S, V]        =
+    AggregationFeature.minBy[S, O, V](o)(v)
+  def sum[V : Monoid](v: S => V): Aggregator[S, V, V]                          =
+    Aggregator.prepareMonoid(v)
+  def uniqueCountBy[T](f : S => T): Aggregator[S, Set[T], Int]                 =
+    AggregationFeature.uniqueCountBy(f)
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: No need to vertically align `=` when lines are wrapped like this.

File: core/src/test/scala/commbank/coppersmith/FeatureBuilderSpec.scala [hunk 8]
```diff
@@ -189,13 +218,52 @@ object AggregationFeatureSetSpec extends Specification with ScalaCheck { def is
...
+    case FloatingPoint(Some(expectedDouble)) =>
       beCloseTo(expectedDouble +/- 0.000000000001) ^^ (
         (v: Value) => v match {
-          case Decimal(Some(actualDouble)) => actualDouble
+          case FloatingPoint(Some(actualDouble)) => actualDouble
           case _ => Double.NaN
         }
       )
+    case Decimal(Some(expectedBigDecimal)) =>
+      beCloseTo(expectedBigDecimal +/- 0.000000000001) ^^ (
```
Comment (dkristian) [inline core/src/test/scala/commbank/coppersmith/FeatureBuilderSpec.scala]: I'd argue that this should be more precise than `FloatingPoint`, possibly exactly the same even, given that `BigDecimal` is being made available for more accuracy.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 9]
```diff
@@ -112,31 +120,111 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
+          // Handle zero without allocation
+          big
+        } else if (newCnt == 0L) {
+          zero
+        } else {
+          val an = big.value
+          val ak = small.value
+          val scaling = k.toDouble / newCnt
+          // a_n + (a_k - a_n)*(k/(n+k)) is only stable if n is not approximately k
+          val newAve = if (scaling < STABILITY_CONSTANT) (an + (ak - an) * scaling) else (n * an + k * ak) / newCnt
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: As mentioned elsewhere, I'm not sure that this is applicable for `BigDecimal`s; at least, not necessarily with the same `STABILITY_CONSTANT` (that "constant" may depend on the individual precisions of the `BigDecimal` values). I would feel more comfortable including this if there were explicit tests around it, both property based, as well as some extreme cases (averaging numbers of vastly different orders of magnitude, as well as `count`s of various sizes that exercise both sides of the `scaling` wrt the constant). Unfortunately there don't appear to be any corresponding tests in algebird to borrow from.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 10]
```diff
@@ -112,31 +120,99 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
+  def max[T, V : Ordering](v: T => V): Aggregator[T, V, V]        = Aggregator.max[V].composePrepare[T](v)
+  def min[T, V : Ordering](v: T => V): Aggregator[T, V, V]        = Aggregator.min[V].composePrepare[T](v)
+  def uniqueCountBy[S, T](f : S => T): Aggregator[S, Set[T], Int] = Aggregator.uniqueCount[T].composePrepare(f)
 
   // TODO: Would be surprised if this doesn't exist elsewhere
   implicit class AlgebirdSemigroup[T](s: Semigroup[T]) {
     def toScalaz = new scalaz.Semigroup[T] { def append(t1: T, t2: =>T): T = s.plus(t1, t2) }
   }
+
+  // Duplicate code from com.twitter.algebird to allow BigDecimals
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: Needs a quick comment about where and why it differs.

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala [hunk 11]
```diff
@@ -112,31 +120,99 @@ trait AggregationFeatureSet[S] extends FeatureSet[(EntityId, Iterable[S])] {
...
+          zero
+        } else {
+          val an = big.value
+          val ak = small.value
+          val newAve = (n * an + k * ak) / newCnt
+          new BigDAveragedValue(newCnt, newAve)
+        }
+      }
+    }
+
```
Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: Minor: additional vertical space

File: core/src/test/scala/commbank/coppersmith/AggregationSpec.scala [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    val aggregationResults = bigDAvgValLists.map(plus)
+    val actual = plus(aggregationResults)
+
+    actual.value must beCloseTo(expected.value +/- BigDecimal("0.0000000000000000001"))
+  }}
+
+  def plus(vals: NonEmptyList[BigDAveragedValue]) = {
+    vals.tail.foldLeft(vals.head)((l, r) => BigDAveragedGroup.plus(l, r))
+  }
+
```
Comment (dkristian) [inline core/src/test/scala/commbank/coppersmith/AggregationSpec.scala]: Minor: additional vertical space

**Comment by dkristian** [comment]
A couple of minor comments, otherwise :+1:

File: core/src/main/scala/commbank/coppersmith/FeatureSet.scala (hunk 10 above)
Comment (ewanv) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: Is
```
// Duplicate code from com.twitter.algebird to allow BigDecimals  
// The scaling code in com.twitter.algebird.AveragedGroup.plus has been removed
// as it creates inaccuracy when dealing with BigDecimals
```
suitable?

Comment (dkristian) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: How about
```scala
// Based on com.twitter.algebird.AveragedGroup. Omits the scaling code
// required for increasing the accuracy of averaging `Double`s, as it is
// not applicable for the `BigDecimal` implementation.
```

Comment (ewanv) [inline core/src/main/scala/commbank/coppersmith/FeatureSet.scala]: Looks good

**Pull request merged**